// @taroify/core/calendar/calendar.shared.js
export { getEndDayOfMonth } from "../datetime-picker/datetime-picker.shared.js";
export function compareYearMonth(date1, date2) {
    const year1 = date1?.getFullYear();
    const year2 = date2?.getFullYear();
    if (year1 === year2) {
        const month1 = date1?.getMonth();
        const month2 = date2?.getMonth();
        return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
    }
    return year1 > year2 ? 1 : -1;
}
export function compareDate(day1, day2) {
    const compareMonthResult = compareYearMonth(day1, day2);
    if (compareMonthResult === 0) {
        const date1 = day1?.getDate();
        const date2 = day2?.getDate();
        return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
    }
    return compareMonthResult;
}
export const cloneDate = (date) => new Date(date);
export function createDayByOffset(date, offset) {
    const cloned = cloneDate(date);
    cloned.setDate(cloned.getDate() + offset);
    return cloned;
}
export const createPreviousDay = (date) => createDayByOffset(date, -1);
export const createNextDay = (date) => createDayByOffset(date, 1);
export function createToday() {
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    return today;
}
export const MIN_DATE = createToday();
export const MAX_DATE = new Date(MIN_DATE.getFullYear(), MIN_DATE.getMonth() + 6, 14);
export const genMonthId = (date) => `taroify-calendar-${date.getFullYear()}-${date.getMonth()}`;
