// @taroify/core/transition/transition.js
import * as React from "react";
import { cloneElement, isValidElement, useMemo, useRef, useState } from "react";
import { CSSTransition } from "react-transition-group";
import { prefixClassname } from "../styles/index.js";
export var TransitionName;
(function (TransitionName) {
    TransitionName["Fade"] = "fade";
    TransitionName["SlideUp"] = "slide-up";
    TransitionName["SlideDown"] = "slide-down";
    TransitionName["SlideLeft"] = "slide-left";
    TransitionName["SlideRight"] = "slide-right";
})(TransitionName || (TransitionName = {}));
const TRANSITION_PRESETS = [
    TransitionName.Fade,
    TransitionName.SlideUp,
    TransitionName.SlideDown,
    TransitionName.SlideLeft,
    TransitionName.SlideRight,
];
function isTransitionPreset(name) {
    return name && TRANSITION_PRESETS.includes(name);
}
function useElementStyle(children) {
    return useMemo(() => {
        if (!isValidElement(children)) {
            return {};
        }
        const element = children;
        const { style } = element.props;
        return style ?? {};
    }, [children]);
}
export default function Transition(props) {
    const { name, in: inProp = false, appear = false, mountOnEnter = false, unmountOnExit, timeout = 300, children: childrenProp, onEnter, onEntering, onEntered, onExit, onExiting, onExited, } = props;
    const children = useMemo(() => childrenProp, [childrenProp]);
    const childrenStyle = useElementStyle(children);
    const transactionName = isTransitionPreset(name) ? prefixClassname(`transition-${name}`) : name;
    const [enter, setEnter] = useState(false);
    const [exited, setExited] = useState(false);
    const ref = useRef(null);
    const setRef = (value) => {
        ref.current = value;
    };
    return (React.createElement(CSSTransition, { in: inProp, nodeRef: ref, mountOnEnter: mountOnEnter, unmountOnExit: unmountOnExit, timeout: timeout, appear: appear, classNames: transactionName, style: {
            ...childrenStyle,
            display: enter && !exited ? "" : "none",
        }, onEnter: (node, isAppearing) => {
            setEnter(true);
            setExited(false);
            onEnter?.(node, isAppearing);
        }, onEntering: onEntering, onEntered: onEntered, onExit: onExit, onExiting: onExiting, onExited: (node) => {
            setEnter(false);
            setExited(true);
            onExited?.(node);
        } }, cloneElement(children, { ref: setRef })));
}
